src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -66,50 +112,133 @@ public void provide(DependencyInfo dependency, Project project, LoomGradleExtens
+		} else {
+			// These are merged v1 mappings
+			saveMergedV1Mappings(project, yarnJar.toPath());
+
+			if (tinyMappings.exists()) {
+				tinyMappings.delete();
 			}
 
 			project.getLogger().lifecycle(":populating field names");
-			new CommandProposeFieldNames().run(new String[]{minecraftProvider.MINECRAFT_MERGED_JAR.getAbsolutePath(), MAPPINGS_TINY_BASE.getAbsolutePath(), MAPPINGS_TINY.getAbsolutePath()});
+			suggestFieldNames(minecraftProvider, tinyMappingsWithoutEnums, tinyMappings.toPath());
 		}
+	}
 
-		mappedProvider = new MinecraftMappedProvider();
-		mappedProvider.initFiles(project, minecraftProvider, this);
-		mappedProvider.provide(dependency, project, extension, postPopulationScheduler);
+	private void saveMergedV1Mappings(Project project, Path mappingsJar) throws IOException {
+		project.getLogger().lifecycle(":extracting " + mappingsJar.getFileName());
+
+		try (FileSystem fileSystem = FileSystems.newFileSystem(mappingsJar, null)) {
+			extractMappings(fileSystem, tinyMappingsWithoutEnums);
+		}
+	}
+
+	private void extractMappings(FileSystem jar, Path extractTo) throws IOException {
+		Files.copy(jar.getPath("mappings/mappings.tiny"), extractTo, StandardCopyOption.REPLACE_EXISTING);
+	}
+
+	private void mergeAndSaveMappings(Project project, Path unmergedIntermediaryJar, Path unmergedYarnJar) throws IOException {
+		Path unmergedIntermediary = Paths.get(mappingsStepsDir.toString(), "unmerged-intermediary.tiny");
+		project.getLogger().lifecycle(":extracting " + unmergedIntermediaryJar.getFileName().toString());
+
+		try (FileSystem unmergedIntermediaryFs = FileSystems.newFileSystem(unmergedIntermediaryJar, null)) {
+			extractMappings(unmergedIntermediaryFs, unmergedIntermediary);
+		}
+
+		Path unmergedYarn = Paths.get(mappingsStepsDir.toString(), "unmerged-yarn.tiny");
+		project.getLogger().lifecycle(":extracting " + unmergedYarnJar.getFileName().toString());
+
+		try (FileSystem unmergedYarnJarFs = FileSystems.newFileSystem(unmergedYarnJar, null)) {
+			extractMappings(unmergedYarnJarFs, unmergedYarn);
+		}
+
+		Path invertedIntermediary = Paths.get(mappingsStepsDir.toString(), "inverted-intermediary.tiny");
+		reorderMappings(unmergedIntermediary, invertedIntermediary, "intermediary", "official");
+		Path unorderedMergedMappings = Paths.get(mappingsStepsDir.toString(), "unordered-merged.tiny");
+		project.getLogger().lifecycle(":merging");
+		mergeMappings(invertedIntermediary, unmergedYarn, unorderedMergedMappings);
+		reorderMappings(unorderedMergedMappings, tinyMappings.toPath(), "official", "intermediary", "named");
+	}
+
+	private void reorderMappings(Path oldMappings, Path newMappings, String... newOrder) {
+		Command command = new CommandReorderTinyV2();
+		String[] args = new String[2 + newOrder.length];
+		args[0] = oldMappings.toAbsolutePath().toString();
+		args[1] = newMappings.toAbsolutePath().toString();
+		System.arraycopy(newOrder, 0, args, 2, newOrder.length);
```

Comment:
Minor thing: Possibly use ArrayUtils.addAll to combine the 2 arrays?

```java
	private void reorderMappings(java.nio.file.Path oldMappings, Path newMappings, String... newOrder) {
		Command command = new CommandReorderTinyV2();
		String[] args = new String[]{oldMappings.toAbsolutePath().toString(), newMappings.toAbsolutePath().toString()};
		runCommand(command, org.apache.commons.lang3.ArrayUtils.addAll(args, newOrder));
	}
```

Something like I posted just seems a bit easier to read, and possibly less prone to bugs if someone comes along to touch this later down the line.

Reply:
we don't actually have lang3 outside of internal gradle api

Reply:
Ok. leave this. Was just a minor thing as well